src/context/Context.tsx

```diff
@@ -59,48 +36,68 @@ function MoviesShowsProvider({ children }: ContextProps) {
     // The search is performed only when there are 3 or more characters in the search bar
     // It should be triggered only one second after the user has stopped typing
     if (queryType === QUERY_TYPE.SEARCH) {
-      timer = setTimeout(
-        () =>
-          getItems(queryType).then(() => {
-            if (timer) {
-              clearTimeout(timer);
-              timer = null;
-            }
-          }),
-        DELAY,
-      );
+      timer = setTimeout(() => getItemsDataAndClearTimer(queryType), DELAY);
     } // Prevent getting top 10 items multiple times if there are 2 or less characters in the search bar
     else if (queryType !== activeQueryType) {
-      getItems(queryType);
+      getItemsData(queryType);
     }
-  }, [search, activeQueryType]); // eslint-disable-line react-hooks/exhaustive-deps
+  }, [search, activeQueryType]);
 
   // Triggered on switching between tabs
   useEffect(() => {
-    getItems(activeQueryType);
-  }, [contentType]); // eslint-disable-line react-hooks/exhaustive-deps
+    getItemsData(activeQueryType);
+  }, [contentType]);
+
+  const getItemsData = (queryType: string) =>
+    getItems(queryType, contentType, search)
+      .then(({ results }) => {
+        const items = queryType === QUERY_TYPE.TOP_RATED ? results.slice(0, NUMBER_OF_ITEMS) : results;
+
+        if (contentType === CONTENT_TYPE.TV_SHOW) {
+          setShows(items);
+        } else {
+          setMovies(items);
+        }
+      })
+      .catch(() => {
+        // TODO: Handle errors
+      })
+      .finally(() => {
+        setLoading(false);
+      });
+
+  const getItemsDataAndClearTimer = (queryType: string) => {
+    getItemsData(queryType).then(() => {
```

Comment:
What will happen if the API request fails? Is **then** callback going to be executed?

Reply:
`then` callback is not going to be executed if the API request fails. I forgot to add `catch` part.